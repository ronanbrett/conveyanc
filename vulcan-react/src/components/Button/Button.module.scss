@import "src/styles/vars";

.Button {
  border: none;
  background-color: var(--highlight-color);
  padding: var(--spacing-s) var(--spacing-m);
  border: 2px solid white;

  transition: all 1s cubic-bezier(0.075, 0.82, 0.165, 1);

  border-radius: 8px;
  cursor: pointer;

  &:focus {
    outline: none;
  }

  &:hover {
    border: 2px solid var(--highlight-color);
    background-color: white;
  }
}
