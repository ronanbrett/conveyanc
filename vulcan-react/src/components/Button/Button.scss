@import "src/styles/vars";

.Button {
  border: none;
  padding: var(--spacing-s) var(--spacing-m);
  border: 2px solid white;

  transition: border 1s cubic-bezier(0.075, 0.82, 0.165, 1);
  transition: background 1s cubic-bezier(0.075, 0.82, 0.165, 1);

  border-radius: 8px;
  cursor: pointer;

  &:focus {
    outline: none;
  }

  &--primary {
    background-color: var(--highlight-color);
    border: 2px solid white;

    &:hover {
      border: 2px solid var(--highlight-color);
      background-color: white;
    }
  }

  &--outline {
    background-color: transparent;
    border: 2px solid var(--highlight-color);
    font-weight: 400;

    &:hover {
      border: 2px solid var(--highlight-color);
    }
  }

  &--small {
    padding: var(--spacing-s) var(--spacing-m);
  }

  &--large {
    padding: var(--spacing-m) var(--spacing-l);
  }
}
