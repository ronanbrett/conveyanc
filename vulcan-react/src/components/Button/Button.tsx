import React, { forwardRef, SyntheticEvent, useState } from "react";
import styles from "./Button.module.scss";

interface ButtonProps {
  a11yTitle?: any;

  label?: string;
  onClick?: (event?: any) => void;
  children?: any;
  [id: string]: any;
}

const Button = forwardRef(
  (
    {
      a11yTitle,
      active,
      align = "center",
      color, // munged to avoid styled-components putting it in the DOM
      children,
      disabled,
      icon,
      focusIndicator = true,
      gap = "small",
      fill, // munged to avoid styled-components putting it in the DOM
      href,
      kind: kindArg,
      label,
      onBlur,
      onClick,
      onFocus,
      onMouseOut,
      onMouseOver,
      plain,
      primary,
      reverse,
      secondary,
      selected,
      size,
      type = "button",
      as,
      ...rest
    }: Partial<ButtonProps>,
    ref
  ) => {
    const [focus, setFocus] = useState<boolean>();
    const [hover, setHover] = useState<boolean>(false);

    const onMouseOverButton = (event: SyntheticEvent) => {
      setHover(true);
      if (onMouseOver) {
        onMouseOver(event);
      }
    };

    const onMouseOutButton = (event: SyntheticEvent) => {
      setHover(false);
      if (onMouseOut) {
        onMouseOut(event);
      }
    };

    const domTag = !as && href ? "a" : as;

    if ((icon || label) && children) {
      console.warn(
        "Button should not have children if icon or label is provided"
      );
    }

    const wrappedClasses = `${styles.Button}  ${active ? "active" : ""} ${
      focus ? "focused" : ""
    } ${hover ? "hover" : ""}`.trim();

    return (
      <button
        className={`${wrappedClasses}`}
        onClick={onClick}
        ref={ref as any}
        aria-label={a11yTitle}
        disabled={disabled}
        onFocus={(event) => {
          setFocus(true);
          if (onFocus) {
            onFocus(event);
          }
        }}
        onBlur={(event) => {
          setFocus(false);
          if (onBlur) {
            onBlur(event);
          }
        }}
        onMouseOver={onMouseOverButton}
        onMouseOut={onMouseOutButton}
        type={!href ? type : undefined}
        {...rest}
      >
        {children}
      </button>
    );
  }
);

// const Button = forwardRef(({ children, props }, ref));
// // const Button: FC<ButtonProps> = ({ children, ...props }) => {
// //   return (
// //     <button className={styles.Button} {...props}>
// //       {children}
// //     </button>
// //   );
// // };

export default Button;
