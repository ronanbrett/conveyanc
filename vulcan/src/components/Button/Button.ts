import { defineComponent } from 'vue';

const Button = defineComponent({});

export default Button;
