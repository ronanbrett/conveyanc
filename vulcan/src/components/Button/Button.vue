<template>
  <button @click="this.$emit('click', event)">
    <slot></slot>
  </button>
</template>

<script src="./Button.ts" lang="ts"></script>

<!-- Add "scoped" attribute to limit SCSS to this component only -->
<style scoped lang="scss">
@import './src/styles/vars';

button {
  border: none;
  background-color: var(--highlight-color);
  padding: var(--spacing-s) var(--spacing-m);
  border: 2px solid white;

  transition: all 1s cubic-bezier(0.075, 0.82, 0.165, 1);

  border-radius: 8px;
  cursor: pointer;

  &:hover {
    border: 2px solid var(--highlight-color);
    background-color: white;
  }
}
</style>
