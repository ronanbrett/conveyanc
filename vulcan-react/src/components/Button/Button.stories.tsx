import React from "react";

import Button from "./Button";

import { action } from "@storybook/addon-actions";

export default {
  title: "Components/Button",
  decorators: [],
  component: Button,
};

export const Default = (props) => <Button {...props}>I am a Button</Button>;

export const Outline = (props) => (
  <Button buttonType="outline" {...props}>
    I am a Button
  </Button>
);

export const Small = (props) => (
  <Button size="small" buttonType="outline" {...props}>
    I am a Button
  </Button>
);
