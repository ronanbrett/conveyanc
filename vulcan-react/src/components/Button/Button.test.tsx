import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Button from "./Button";
const { axe, toHaveNoViolations } = require("jest-axe");
expect.extend(toHaveNoViolations);

describe("Button", () => {
  it("renders without error", () => {
    const { asFragment } = render(<Button />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should be compliant", async () => {
    const { container } = render(<Button />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it("should be able to trigger change events", () => {
    const mockCallBack = jest.fn();

    const { container } = render(<Button onClick={mockCallBack} />);
    container.querySelector("button").click();
    expect(mockCallBack).toHaveBeenCalled();
  });

  it("should set hover on hover", () => {
    const { container } = render(<Button />);
    let button = container.querySelector("button");

    fireEvent.mouseOver(button);

    expect(button).toHaveClass("hover");
  });

  it("should set focus on hover", () => {
    const { container } = render(<Button />);
    let button = container.querySelector("button");

    fireEvent.focus(button);

    expect(button).toHaveClass("focus");
  });

  it("should not emit events if its disabled", () => {
    const mockCallBack = jest.fn();

    const { container } = render(<Button onClick={mockCallBack} disabled />);
    let button = container.querySelector("button");

    fireEvent.click(button);

    expect(mockCallBack).not.toHaveBeenCalled();
  });
});
