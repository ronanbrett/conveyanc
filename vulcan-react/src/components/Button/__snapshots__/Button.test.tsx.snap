// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders without error 1`] = `
<DocumentFragment>
  <button
    class="Button"
    type="button"
  />
</DocumentFragment>
`;
